workflow/controller/operator.go

```diff
@@ -1144,11 +1144,19 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) (error, bool)
 		node, err := woc.wf.Status.Nodes.Get(nodeID)
 		if err == nil {
 			if newState := woc.assessNodeStatus(ctx, pod, node); newState != nil {
-				// Check whether its taskresult is in an incompleted state.
-				if newState.Succeeded() && woc.wf.Status.IsTaskResultIncomplete(node.ID) {
-					woc.log.WithFields(log.Fields{"nodeID": newState.ID}).Debug("Taskresult of the node not yet completed")
-					taskResultIncomplete = true
-					return
+				// Check whether its taskResult is in an incomplete state.
+				if woc.wf.Status.IsTaskResultIncomplete(node.ID) {
+					woc.log.WithFields(log.Fields{"nodeID": newState.ID}).Debug("TaskResult of the node not yet completed")
+					switch {
+					case newState.Succeeded():
+						woc.log.WithFields(log.Fields{"nodeID": newState.ID}).Debug("Node succeeded, waiting for its taskResult to be completed")
+						taskResultIncomplete = true
+						return
+					case newState.FailedOrError():
+						// The wait container might have been terminated not gracefully, mark its taskResult as completed here.
+						woc.log.WithFields(log.Fields{"nodeID": newState.ID}).Debug("Node failed or error, mark its taskResult as completed")
+						woc.wf.Status.MarkTaskResultComplete(node.ID)
```

Comment:
At this point the task result might have been written by the wait container, but the workflow controller might not have observed this. Just because the pod has failed it doesn't **always** mean the wait container didn't get a chance to write a workflowTaskResult.

Shouldn't we wait for the workflowTaskResult to possibly come through and check for this state somewhere around [here](https://github.com/argoproj/argo-workflows/blob/dcd943618e4e85443c75944f5f2c333f07c89805/workflow/controller/taskresult.go#L99). We don't record the transition time into failed, so this may be tricky to implement.

Reply:
https://github.com/argoproj/argo-workflows/blob/dcd943618e4e85443c75944f5f2c333f07c89805/workflow/controller/operator.go#L1498-L1505

`assessNodeStatus` will check the status of `wait` container, so if the new state is failed or error, it also means `wait` container has been terminated.

Reply:
This does make sense. There is nothing to say that under these circumstances we might eventually receive a completed WorkflowTaskResult, although it does seem the more unlikely scenario.

The delay introduced in `podAbsentTimeout` is for similar reasons to the question above. At least here we've got positive confirmation of pod death, but we don't **know** that a completed WorkflowTaskResult hasn't yet arrived, we're assuming it won't.

The story of WorkflowTaskResults has changed during 3.5. Of note is #12574 which made WorkflowTaskResults completion a requirement.

#13454 and this PR are both trying to partially undo that requirement - for good reasons. What I'm trying to ensure is that this isn't an endless spiral of adding complexity circling around this issue.

I'm happy to merge it as is, because it's better than before. It's limited to pods that have been Evicted now, which probably means there are more pods that can slip through the cracks.

I think we'll also need a patch to convert [this](https://github.com/argoproj/argo-workflows/blob/dcd943618e4e85443c75944f5f2c333f07c89805/workflow/controller/taskresult.go#L99) test from
```go
if !foundPod && !node.Completed() {
```
to something like
```go
if (!foundPod || pod.Completed()) && !node.Completed() {
```
to cover my other concerns, but that can happen another day.